### PR TITLE
fastapi: first milestone

### LINKED ISSUE #89: include Lotus instance in docker-compose
This would be nice to make running compose _complete_, and also avoid using _network_ of type _host_ to being able to connect to Lotus running on the host.

### LINKED ISSUE #110: fpa: Create instance manager
Create n FastAPI instance manager which will own created instances. Maps authentication tokens with instances, and will have all the logic regarding token invalidation, permissions, and life-cycle management.

### LINKED ISSUE #112: fpa: docker compose
Will be useful to have a docker-compose for trying implementation. Would include the server, ipfs, and something that at least simulates reasonably well FIL network since testnet isn't viable (evaluate at the moment).

### LINKED ISSUE #119: fpa: Get and Store with car transform

### LINKED ISSUE #120: fpa: instance visibility apis
A bunch of APIs to see config, tracking history, id, wallet balance.

### LINKED ISSUE #121: use PARAMCACHE_PATH
It might be better to make running tests less invasive to a possibly running daemon.
This case is already covered in the Makefile, but would be cleaner to do.

### LINKED ISSUE #126: fil-toolsd with embedded mocked network
Allow running a _fil-toolsd_ instance with an embedded FIL network. This would allow to use the server without depending on a real lotus node.

### LINKED ISSUE #127: wallet: allow creating with funds
Allow configuring a master addr and value to be used for funding new created wallets.

### LINKED ISSUE #128: fpa: gRPC service

### LINKED ISSUE #129: fpa: CLI using gRPC client

### LINKED ISSUE #130: fpa: miner selection for deals
Create an abstraction which can be implemented with different strategies.

### PR DESCRIPTION
**Pull request by jsign**
Closes #121 
Closes #110 
Closes #89 
Closes #112 
Closes #126 
Closes #127 
Closes #129 
Closes #128 
Closes #119
Closes #120 
Closes #130

Some code comments to help... this PR got a bit massive for my original intentions.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -9,19 +9,19 @@ on:
 jobs:
   test:
     name: Test
-    runs-on: self-hosted
+    runs-on: ubuntu-latest
```
Comment (jsign) [inline .github/workflows/test.yml]: Returning to _self-hosted_ mode.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -9,19 +9,19 @@ on:
...
     steps:
       - name: Set up Go
-        uses: actions/setup-go@v1
+        uses: actions/setup-go@v2-beta
```
Comment (jsign) [inline .github/workflows/test.yml]: Let's start using v2.

File: Makefile [hunk 3]
```diff
@@ -20,11 +20,13 @@ build: .update-modules .filecoin-build
 clean:
 	rm -f .filecoin-build
 	rm -f .update-modules
+	git submodule deinit --all -f
```
Comment (jsign) [inline Makefile]: Allow `clean` delete the `git-submodule` and start cleaner.

File: Makefile [hunk 4]
```diff
@@ -20,11 +20,13 @@ build: .update-modules .filecoin-build
...
 
+PARAMCACHE_PATH:=/var/tmp/fil-tools/filecoin-proof-parameters
```
Comment (jsign) [inline Makefile]: #121

File: api/client/fpa.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
+package client
```
Comment (jsign) [inline api/client/fpa.go]: #129

File: api/server/server.go [hunk 6]
```diff
@@ -70,8 +79,11 @@ type Server struct {
 
 // Config specifies server settings.
 type Config struct {
+	WalletInitialFunds  big.Int
```
Comment (jsign) [inline api/server/server.go]: #127

File: api/server/server.go [hunk 7]
```diff
@@ -81,55 +93,83 @@ type Config struct {
 
 // NewServer starts and returns a new server with the given configuration.
 func NewServer(conf Config) (*Server, error) {
-	c, cls, err := lotus.New(conf.LotusAddress, conf.LotusAuthToken)
+	var c *apistruct.FullNodeStruct
+	var cls func()
+	var err error
+	var masterAddr address.Address
+	if conf.Embedded {
```
Comment (jsign) [inline api/server/server.go]: #126

File: api/server/server.go [hunk 8]
```diff
@@ -81,55 +93,83 @@ type Config struct {
...
 	if err != nil {
-		return nil, fmt.Errorf("error when creating slashing index: %s", err)
+		return nil, fmt.Errorf("creating slashing index: %s", err)
 	}
 	dm, err := deals.New(c, deals.WithImportPath(filepath.Join(conf.RepoPath, "imports")))
 	if err != nil {
-		return nil, fmt.Errorf("error when creating deal module: %s", err)
+		return nil, fmt.Errorf("creating deal module: %s", err)
+	}
+	wm, err := wallet.New(c, &masterAddr, conf.WalletInitialFunds)
```
Comment (jsign) [inline api/server/server.go]: The Wallet Module is configured now with a _masterAddr_ and a _WalletInitialFunds_.
Every new wallet created with be sent that amount of _attoFIL_ amount from _masterAddr_.
If _masterAddr_ is nil, then no funds will be sent.

Most probably this will change a bit more in the future since it may be better to configure this action _per wallet creation_ and not globally.

File: api/server/server.go [hunk 9]
```diff
@@ -81,55 +93,83 @@ type Config struct {
...
-	wm := wallet.New(c)
 	rm := reputation.New(txndstr.Wrap(ds, "reputation"), mi, si, ai)
 
+	ipfs, err := httpapi.NewApi(conf.IpfsApiAddr)
+	if err != nil {
+		return nil, fmt.Errorf("creating ipfs client: %s", err)
+	}
+
+	reptop := reptop.New(rm, ai)
+	fpamanager, err := fpa.New(txndstr.Wrap(ds, "fpa"), wm, dm, reptop, &noopauditer.Auditer{}, ipfs)
```
Comment (jsign) [inline api/server/server.go]: FastAPI manager.

File: chainstore/chainstore.go [hunk 10]
```diff
@@ -144,15 +144,12 @@ func (s *Store) save(ts types.TipSetKey, state interface{}) error {
...
 	c := checkpoint{id: s.lastID + 1, ts: ts}
-	if err := s.ds.Put(toKeyData(c.ts), buf); err != nil {
+	if err := txn.Put(toKeyData(c.ts), buf); err != nil {
 		return err
 	}
-	if err := s.ds.Put(toKeyID(c.id), c.ts.Bytes()); err != nil {
-		return nil
+	if err := txn.Put(toKeyID(c.id), c.ts.Bytes()); err != nil {
+		return err
 	}
```
Comment (jsign) [inline chainstore/chainstore.go]: Bugfix. A txn was created, but actions weren't using it. Wasn't causing problems since it also works, but wasn't correct.

File: chainsync/chainsync_test.go [hunk 11]
```diff
@@ -6,7 +6,7 @@ import (
 	"testing"
 
 	"github.com/filecoin-project/lotus/chain/types"
-	logging "github.com/ipfs/go-log"
+	logging "github.com/ipfs/go-log/v2"
```
Comment (jsign) [inline chainsync/chainsync_test.go]: Migrate all _go-log_ usage to _v2_.

File: deals/deals.go [hunk 12]
```diff
@@ -63,42 +64,50 @@ func New(api API, opts ...Option) (*Module, error) {
 
 // Store creates a proposal deal for data using wallet addr to all miners indicated
 // by dealConfigs for duration epochs
-func (m *Module) Store(ctx context.Context, waddr string, data io.Reader, dcfgs []StorageDealConfig, dur uint64) (cid.Cid, []cid.Cid, []StorageDealConfig, error) {
+func (m *Module) Store(ctx context.Context, waddr string, data io.Reader, dcfgs []StorageDealConfig, dur uint64) (cid.Cid, []StoreResult, error) {
```
Comment (jsign) [inline deals/deals.go]: Basically, instead of returning _successful_ and _failed_ proposals as different lists, only one list is returned which as an attribute saying if failed or not. Is much more convenient for use.

File: deals/deals_test.go [hunk 13]
```diff
@@ -25,22 +24,22 @@ func TestMain(m *testing.M) {
 }
 
 func TestStore(t *testing.T) {
-	numMiners := []int{1, 3}
+	numMiners := []int{1}
```
Comment (jsign) [inline deals/deals_test.go]: Make tests lighter to free GH CI... the custom runner was also having problems, so better use the free one.
I still have pending looking for another solution to this... in my personal laptop heavier tests run correctly.

File: deals/service.go [hunk 14]
```diff
@@ -4,7 +4,6 @@ import (
 	"context"
```
Comment (jsign) [inline deals/service.go]: All this changes are adjustments to a change in the `deals.Store(..)` api that merged two output lists into a single one (as mentioned in an earlier comment).

File: deals/types.go [hunk 15]
```diff
@@ -3,14 +3,19 @@ package deals
...
 
-	"github.com/filecoin-project/lotus/chain/types"
 	"github.com/ipfs/go-cid"
 )
 
 // StorageDealConfig contains information about a storage proposal for a miner
 type StorageDealConfig struct {
 	Miner      string
-	EpochPrice types.BigInt
+	EpochPrice uint64
```
Comment (jsign) [inline deals/types.go]: Prefer using simpler types for API interfaces. This change scattered small changes in clients of the API, nothing fancy.

File: docker/Makefile [hunk 16]
```diff
@@ -1,13 +1,12 @@
...
-	@TEXFILTOOLS_LOTUSTOKEN=$(LT) docker-compose down
+	docker-compose down
+	docker-compose -p embed down
 .PHONY: down
 
-up: 
-	TEXFILTOOLS_LOTUSTOKEN=$(LT) docker-compose up --build 
-.PHONY: up
\ No newline at end of file
+fresh: down
```
Comment (jsign) [inline docker/Makefile]: Simplification of makefile. All env variables are defined in the _docker-compose_ files.

File: docker/Makefile [hunk 17]
```diff
@@ -1,13 +1,12 @@
...
+	docker-compose up --build 
+.PHONY: fresh-up
+
+embed:
```
Comment (jsign) [inline docker/Makefile]: An extra rule for running in _embedded_ mode.

File: docker/docker-compose.yaml [hunk 18]
```diff
@@ -2,24 +2,52 @@ version: '3.7'
 
 volumes:
   textile-fc-data:
+  textile-fc-ipfs:
+  textile-fc-lotus:
 
 services:
-  textile-fc:
-    network_mode: host
+
```
Comment (jsign) [inline docker/docker-compose.yaml]: Since now this _docker-compose_ file also includes the Lotus node, we don't have the problem of _fil-tools_ container needing to access a host port (where lotus was running). This means we don't need fancy _network-modes_ since _fil-tools_ now communicates to another container in the same docker-compose environment.

This is good since _network_mode=host_ was only supported in Linux... so now this can be run in Darwin too.

File: docker/docker-compose.yaml [hunk 19]
```diff
@@ -2,24 +2,52 @@ version: '3.7'
...
+      - lotus
+    environment:
+      - TEXFILTOOLS_LOTUSHOST=/dns4/lotus/tcp/1234
+      - TEXFILTOOLS_IPFSAPIADDR=/dns4/ipfs/tcp/5001
+      - TEXFILTOOLS_LOTUSTOKENFILE=/root/lotus/.lotus/token
+      - TEXFILTOOLS_REPOPATH=/root/fil-tools/.texfc
+    restart: unless-stopped
+    volumes:
+      - textile-fc-data:/root/fil-tools
+      - textile-fc-lotus:/root/lotus
```
Comment (jsign) [inline docker/docker-compose.yaml]: Leveraging docker-volumes to get the API-token from the _lotus_ container in the _fil-tools_ container.

File: docker/grafana/provisioning/dashboards/TextileFC.json [hunk 20]
```diff
@@ -184,9 +184,9 @@
       "tableColumn": "",
```
Comment (jsign) [inline docker/grafana/provisioning/dashboards/TextileFC.json]: All _docker/grafana/*_ changes are due to monitoring should now use docker container names instead of _localhost_. (since not running in _network-mode=host_)

Comment (jsign) [reply docker/grafana/provisioning/dashboards/TextileFC.json]: Plus, some name-changes.

File: exe/fil-tools/cmd/fpa.go [hunk 21]
```diff
@@ -0,0 +1,15 @@
+package cmd
```
Comment (jsign) [inline exe/fil-tools/cmd/fpa.go]: #129

File: exe/fil-tools/cmd/root.go [hunk 22]
```diff
@@ -85,3 +88,22 @@ func initConfig() {
 		fmt.Println("Using config file:", viper.ConfigFileUsed())
 	}
 }
+
```
Comment (jsign) [inline exe/fil-tools/cmd/root.go]: Interceptor for auth-token to context handling.

File: exe/server/Dockerfile [hunk 23]
```diff
@@ -1,11 +1,21 @@
 FROM golang:1.13.7 as builder
```
Comment (jsign) [inline exe/server/Dockerfile]: Make the Dockerfile of _fil-tools_ better. 
Instal _rustc_ for Linux and not use the copied compiled things from the host (in the `COPY . .` line).

Realized about this problem when I changed to macOS, since the compiled rust git-submodule was for darwin arch, and the container is Linux. This will make a completely clean build of fil-tools in the docker image

File: exe/server/main.go [hunk 24]
```diff
@@ -106,12 +124,16 @@ func setupInstrumentation() {
...
 func setupLogging() {
-	logging.SetDebugLogging()
-	if !config.GetBool("debug") {
-		logging.SetLogLevel("*", "info")
-		logging.SetLogLevel("rpc", "error")
-		logging.SetLogLevel("dht", "error")
-		logging.SetLogLevel("swarm2", "error")
+	logging.SetLogLevel("*", "error")
+	loggers := []string{"index-miner", "index-ask", "index-slashing", "server",
+		"deals", "fil-toolsd", "fchost", "fpaauth", "fastapi", "ip2location", "reputation"}
```
Comment (jsign) [inline exe/server/main.go]: Not convincing yet. Created #141 since this is a static assignment... future modules can be forgot to add here.

File: fpa/auth/repo.go [hunk 25]
```diff
@@ -0,0 +1,77 @@
+package auth
```
Comment (jsign) [inline fpa/auth/repo.go]: _fps/auth_ is a package that has the information about which _auth-tokens_ for a FastAPI instance maps with the FastAPI instance ID. In the future any kind of _auth-token_ invalidation, expiry, or things related should live here.
TL;DR: mapping between auth-tokens to FastAPI instances.

File: fpa/fastapi/fastapi.go [hunk 26]
```diff
@@ -0,0 +1,209 @@
+package fastapi
```
Comment (jsign) [inline fpa/fastapi/fastapi.go]: The _fastapi_ package is partitioned in three files:
- _fastapi.go_ for general info
- _fastapi_get.go_ for retrieval logic of CIDs
- _fastapi_put.go_ for pinning CIDs.

File: fpa/jsonauditer/jsonauditer.go [hunk 27]
```diff
@@ -0,0 +1,62 @@
+package jsonauditer
```
Comment (jsign) [inline fpa/jsonauditer/jsonauditer.go]: This is a WIP.

The `jsonauditer` or `noopauditer` are implementations of an `Auditer` concept. This `Auditer` is a possible idea to have an event-log of events that happens in a Cid for an Fast API instance.

File: fpa/manager.go [hunk 28]
```diff
@@ -0,0 +1,85 @@
+package fpa
```
Comment (jsign) [inline fpa/manager.go]: The `fpa/manager` is a package that creates/loads Fast API instances.

File: fpa/minerselector/fixed/fixed.go [hunk 29]
```diff
@@ -0,0 +1,33 @@
+package fixed
```
Comment (jsign) [inline fpa/minerselector/fixed/fixed.go]: The `fpa/minerselector` concept are possible implementations of miner-selection algorithms.

This `fpa/minerselector/fixed` is an implementation that always return a fixed miner.

File: fpa/minerselector/reptop/reptop.go [hunk 30]
```diff
@@ -0,0 +1,43 @@
+package reptop
```
Comment (jsign) [inline fpa/minerselector/reptop/reptop.go]: This `fpa/minerselector/reptop` is a miner-selector algorithm that uses the Reputation Module to select the Top N miners. (`reptop` = _reputation top miners_)

File: fpa/noopauditer/noopauditer.go [hunk 31]
```diff
@@ -0,0 +1,35 @@
+package noopauditer
```
Comment (jsign) [inline fpa/noopauditer/noopauditer.go]: `noopauditer` is an `Auditer` that doesn't save any of the events that are reported for a CID. Mostly for tests at the moment.

File: index/miner/onchain.go [hunk 32]
```diff
@@ -31,6 +31,9 @@ func (mi *MinerIndex) updateOnChainIndex() error {
 	if err != nil {
 		return err
 	}
+	if heaviest.Height()-hOffset <= 0 {
+		return nil
+	}
```
Comment (jsign) [inline index/miner/onchain.go]: Cover an edge case that only happens when the indexes are running when a super-fresh lotus node is syncing from genesis.

File: index/miner/onchain.go [hunk 33]
```diff
@@ -55,9 +58,6 @@ func (mi *MinerIndex) updateOnChainIndex() error {
 	log.Infof("current state height %d, new tipset height %d", chainIndex.LastUpdated, new.Height())
 	if hdiff > fullThreshold || chainIndex.LastUpdated == 0 {
 		mctx, _ = tag.New(mctx, tag.Insert(metricRefreshType, "full"))
-		if err != nil {
-			return err
-		}
```
Comment (jsign) [inline index/miner/onchain.go]: Weird code that most probably was left in a rebase.

File: ldevnet/ldevnet.go [hunk 34]
```diff
@@ -59,9 +60,12 @@ func (ld *LocalDevnet) Close() {
 	ld.closer()
 }
 
-func New(t *testing.T, numMiners int) (*LocalDevnet, error) {
+func New(numMiners int, blockDur time.Duration) (*LocalDevnet, error) {
```
Comment (jsign) [inline ldevnet/ldevnet.go]: Add some extra changes to allow tuning the speed at which the local Testnet generates new blocks.
Also avoid being biased in running on _tests_, since we now support the _embeded mode_ which is running a local devote in a non-testing environment.

File: lotus/client.go [hunk 35]
```diff
@@ -52,6 +54,16 @@ func New(maddr ma.Multiaddr, authToken string) (*apistruct.FullNodeStruct, func(
 
 }
 
+func NewEmbedded() (*apistruct.FullNodeStruct, func(), error) {
+	util.AvgBlockTime = time.Second * 1
```
Comment (jsign) [inline lotus/client.go]: If we're running in embedded mode, tune some configs since blocks aren't produced at the default _AvgBlockTime=30s_ that Mainnet will run but faster (since is a local devnet)

File: reputation/reputation.go [hunk 36]
```diff
@@ -81,6 +82,9 @@ func (rm *Module) AddSource(id string, maddr ma.Multiaddr) error {
 
 // GetTopMiners gets the top n miners with best score
 func (rm *Module) GetTopMiners(n int) ([]MinerScore, error) {
+	if n < 1 {
+		return nil, fmt.Errorf("the number of miners should be greater than zero")
+	}
 	mr := make([]MinerScore, 0, n)
```
Comment (jsign) [inline reputation/reputation.go]: More defensive code to check that `make` won't panic if the client sends a zero or negative value.

File: tests/docker.go [hunk 37]
```diff
@@ -0,0 +1,28 @@
+package tests
```
Comment (jsign) [inline tests/docker.go]: Use of `gitHub.com/ory/dockertest` for running docker containers in tests.

File: wallet/wallet.go [hunk 38]
```diff
@@ -32,14 +39,34 @@ func (m *Module) NewWallet(ctx context.Context, typ string) (string, error) {
...
+			To:       addr,
+			Value:    types.BigInt{Int: m.iAmount},
+			GasLimit: types.NewInt(1000),
+			GasPrice: types.NewInt(0),
+		}
+
+		_, err = m.api.MpoolPushMessage(ctx, msg)
+		if err != nil {
+			return "", fmt.Errorf("transfering funds to new wallet: %s", err)
+		}
```
Comment (jsign) [inline wallet/wallet.go]: Send funds to newly created addresses from the master address.

File: exe/fil-tools/cmd/fpa_get.go [hunk 39]
```diff
@@ -0,0 +1,85 @@
...
+			Fatal(errors.New("get requires token"))
+		}
+		ctx = context.WithValue(ctx, authKey("fpatoken"), token)
+
+		if cidString == "" {
+			Fatal(errors.New("store command needs a cid"))
+		}
+
+		if out == "" {
+			Fatal(errors.New("store command needs an out path to write the data to"))
```
Comment (asutula) [inline exe/fil-tools/cmd/fpa_get.go]: Nit: copy pasta, this isn't the store command

File: exe/server/Dockerfile (hunk 23 above)
Comment (asutula) [reply exe/server/Dockerfile]: So glad you switched to Mac hahah. Although maybe the benefit comes from switching back and forth every once in a while :)

File: fpa/jsonauditer/jsonauditer.go (hunk 27 above)
Comment (asutula) [reply fpa/jsonauditer/jsonauditer.go]: Very cool idea, will be useful for sure. I'd love to display this data in the web ui.

File: fpa/service.go [hunk 40]
```diff
@@ -0,0 +1,164 @@
...
+		return nil, err
+	}
+	return &pb.CreateReply{
+		Id:      id.String(),
+		Address: addr,
+	}, nil
+}
+
+func (s *Service) getInstanceByToken(ctx context.Context) (*fastapi.Instance, error) {
+	token := metautils.ExtractIncoming(ctx).Get("X-fpa-Token")
```
Comment (asutula) [inline fpa/service.go]: Maybe I just missed this, or maybe it's still to be seen below, but how does a client acquire an auth token? Is it only when `Create` is called you get an auth token back? Is there a way to get a new auth token if necessary?

**Comment by asutula** [comment]
That auditor typo is spread all over FYI
